src/Text.tsx

```diff
@@ -1,13 +1,53 @@
+import React from 'react'
 import styled from 'styled-components'
 import {COMMON, SystemCommonProps, SystemTypographyProps, TYPOGRAPHY} from './constants'
+import {ForwardRefComponent as PolymorphicForwardRefComponent} from './utils/polymorphic'
 import sx, {SxProp} from './sx'
-import {ComponentProps} from './utils/types'
 
-const Text = styled.span<SystemTypographyProps & SystemCommonProps & SxProp>`
+const restrictedAs = [
+  'span',
+  'p',
+  'b',
+  'i',
+  'em',
+  'small',
+  'strong',
+  'u',
+  'code',
+  'q',
+  's',
+  'label',
+  'ins',
+  'del',
+  'mark',
+  'kbd',
+  'abbr',
+] as const
+
+const StyledText = styled.span<SystemTypographyProps & SystemCommonProps & SxProp>`
   ${TYPOGRAPHY};
   ${COMMON};
   ${sx};
 `
 
-export type TextProps = ComponentProps<typeof Text>
+export type TextProps = {
+  as?: typeof restrictedAs[number]
+} & React.ComponentProps<'span' & typeof StyledText>
+
+const Text = React.forwardRef(({children, as = 'span', ...rest}, forwardedRef) => {
+  if (__DEV__ && !restrictedAs.includes(as)) {
```

Comment:
The `if (__DEV__)` statement may need to be on its own for the compiler to remove it for production builds. cc @joshblack 

```suggestion
  if (__DEV__) {
    if (!restrictedAs.includes(as)) {
```

Reply:
+1 to the convention of "dev-only" blocks wrapping the code that is stripped way. In either case, I think Terser will optimize away `if (false && condition)` thankfully but I would love it if we could still wrap the block in `if (__DEV__)`

Reply:
Perfect, thanks y'all! I'll adjust this and make note for future cases.